staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/clock/event_clock.go

```diff
@@ -0,0 +1,214 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clock
+
+import (
+	"container/heap"
+	"math/rand"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// EventFunc does some work that needs to be done at or after the
+// given time.  All of the work that needs to be done is finished
+// before this function returns and, if there is an associated
+// `sync.WaitGroup` (see NewFakeEventClock), its counter goes to zero;
+// no other work is left to be completed in goroutines.
+type EventFunc func(time.Time)
+
+// EventClock fires event on time
+type EventClock interface {
+	clock.PassiveClock
+	EventAfterDuration(f EventFunc, d time.Duration)
+	EventAfterTime(f EventFunc, t time.Time)
+}
+
+// RealEventClock fires event on real world time
+type RealEventClock struct {
+	clock.RealClock
+}
+
+// EventAfterDuration schedules an EventFunc
+func (RealEventClock) EventAfterDuration(f EventFunc, d time.Duration) {
+	ch := time.After(d)
+	go func() {
+		select {
+		case t := <-ch:
+			f(t)
+		}
+	}()
+}
+
+// EventAfterTime schedules an EventFunc
+func (r RealEventClock) EventAfterTime(f EventFunc, t time.Time) {
+	now := time.Now()
+	d := t.Sub(now)
+	if d <= 0 {
+		go f(now)
+	} else {
+		r.EventAfterDuration(f, d)
+	}
+}
+
+// FakeEventClock is one whose time does not pass implicitly but
+// rather is explicitly set by invocations of its SetTime method
+type FakeEventClock struct {
+	clock.FakePassiveClock
+
+	// waiters is a heap of waiting work, sorted by time
+	waiters     eventWaiterHeap
+	waitersLock sync.RWMutex
+
+	// clientWG may be nil and if not supplies constraints on time
+	// passing in Run.  The Run method will not pick a new time until
+	// this is nil or its counter is zero.
+	clientWG *sync.WaitGroup
+
+	// fuzz is the amount of noise to add to scheduling.  An event
+	// requested to run at time T will run at some time chosen
+	// uniformly at random from the interval [T, T+fuzz]; the upper
+	// bound is exclusive iff fuzz is non-zero.
+	fuzz time.Duration
+
+	// rand is the random number generator to use in fuzzing
+	rand *rand.Rand
+}
+
+type eventWaiterHeap []eventWaiter
+
+var _ heap.Interface = (*eventWaiterHeap)(nil)
+
+type eventWaiter struct {
+	targetTime time.Time
+	f          EventFunc
+}
+
+// NewFakeEventClock constructor.  The given `r *rand.Rand` must
+// henceforth not be used for any other purpose.  If `r` is nil then a
+// fresh one will be constructed, seeded with the current real time.
+// The clientWG can be `nil` and if not is used to let Run know about
+// additional work that has to complete before time can advance.
+func NewFakeEventClock(t time.Time, clientWG *sync.WaitGroup, fuzz time.Duration, r *rand.Rand) *FakeEventClock {
```

Comment:
Rather than taking a `clientWG *sync.WaitGroup` input, this function should return a `grc GoRoutineCounter` (in addition to the clock).  In particular, the `GoRoutineCounter` should be a `*wgGRC` where `type wgGRC struct {sync.WaitGroup}` (pick any name you like better than `wgGRC`).

Reply:
Updated to fully integrate the wgGRC (i changed this to waitGroupCounter but am open to suggestions) as per the comments here.